### ISSUE TITLE
Evicting KV cache corrupts generation (garbage output once prompt > cache_length)

### ISSUE THREAD

**Issue by mmjerge** [awslabs/keys_values#140]
**Describe the bug**

Autoregressive generation through an **evicting** KV cache (e.g. `lastrec`, `h2o`,
and quantized variants) produces garbage — the model emits EOS/newlines
immediately — whenever the prompt is longer than `cache_length` (so the prefill
spans multiple chunks and eviction occurs). With a cache large enough to hold the
whole prompt (no eviction), generation is correct and matches a plain dense /
no-cache forward. This is a **cliff**, not gradual lossy degradation, and it fires
even when only *irrelevant* tokens are evicted, so it is a mechanism bug rather
than expected sparse-attention information loss.

Since eviction is the whole point of these caches for long context (and eval
normally uses `cache_length < sequence_length`), this affects long-context
generation/eval broadly.

**Evidence (Qwen2.5-0.5B-Instruct, A10G, bf16, eager SDPA)**

Controlled key-value retrieval where the answer is the **last** record (so
`lastrec`, which keeps the most-recent tokens, provably retains the needle *and*
the question), ~2062-token prompts, accuracy = substring match of the value:

| cache_length | eviction? | accuracy |
|---:|:--:|---:|
| 2112 | no  | **0.90** |
| 1900 | yes (drops only first ~162 filler tokens) | **0.00** |
| 1536 | yes | 0.00 |
| 1024 | yes | 0.00 |
| 512  | yes | 0.00 |

- The collapse happens at the *first* eviction, dropping only early filler tokens
  that are irrelevant to a last-record query — inconsistent with legitimate
  lossiness.
- Both `lastrec-default` and `lastrec-torch-quantized8` collapse to 0.00, so it is
  **not** a quantization issue.
- `h2o-torch-quantized8` behaves identically.

**Localization**

Only the first prefill chunk (`input_pos == 0`) takes the `_prefill` path
(attends to itself). Every subsequent chunk — which is where eviction happens —
goes through the `_forward` path and uses `token_positions()` to build the
attention mask (`build_mask_slice` in `keys_values/attention/base.py`). The bug
requires a multi-chunk prefill, so it is in that shared masked-attention /
eviction path. `token_positions()` itself appears to return correct original
positions, which points at the mask-slice / eviction-slot bookkeeping. The
first *generated* token (straight from the prefill final-position logits) is
already wrong, so the corruption is in the prefill, before any decode step.

**Severity is scale-dependent (why it may slip past tests):** a tiny 2-layer
random model in fp32 shows only a mild final-logit perturbation under eviction
(max|Δ| ≈ 0.08, argmax preserved), whereas the real 24-layer model in bf16
collapses to garbage. Unit tests on small models may therefore pass.

**To reproduce**

```python
import torch, lightning as L
from litgpt.prompts import PromptStyle, has_prompt_style, load_prompt_style
from litgpt.tokenizer import Tokenizer
from litgpt.utils import auto_download_checkpoint, load_checkpoint
from keys_values.config import Config
from keys_values.data.constants import LIT_MODEL_FNAME
from keys_values.kvcache.factory import KVCacheFactory, deallocate_kv_cache_buffers_of_model
from keys_values.long_context import LongContextInferenceModel
from keys_values.model import GPT
from keys_values.generate.base import batched_generate_fn
from keys_values.utils import VerbosityLevels

fab = L.Fabric(devices=1, accelerator="cuda", precision="bf16-true")
cd = auto_download_checkpoint(model_name="Qwen/Qwen2.5-0.5B-Instruct")
tok = Tokenizer(cd); cfg = Config.from_file(cd / "model_config.yaml")
ps = load_prompt_style(cd) if has_prompt_style(cd) else PromptStyle.from_config(cfg)
pad = tok.processor.token_to_id("<|endoftext|>") or int(tok.eos_id)
with fab.init_module(empty_init=True): m = GPT(cfg)
load_checkpoint(fab, m, cd / LIT_MODEL_FNAME); m.to(fab.device)

# ~2000-token prompt of "key is value" records; the ANSWER is the LAST record.
import random; rng = random.Random(0)
recs = [f"The value for key '{''.join(rng.choice('abcdefghij') for _ in range(6))}' is {rng.randint(100000,999999)}." for _ in range(180)]
answer = "424242"; recs.append(f"The value for key 'zzzzzz' is {answer}.")
body = ("Records:\n" + " ".join(recs) +
        "\n\nQuestion: what is the value for key 'zzzzzz'? Reply with only the value.\nAnswer:")
ids = tok.encode(ps.apply(body), device=fab.device).unsqueeze(0)
L_ = ids.shape[1]

def gen(cache_len):
    deallocate_kv_cache_buffers_of_model(m)
    m.assign_kv_caches(KVCacheFactory.create(gpt_model=m, name="lastrec-default",
        max_batch_size=1, cache_length=cache_len, dtype=torch.bfloat16))
    m.eval(); m.max_seq_length = L_ + 16
    inf = LongContextInferenceModel(m, head_model=None, chunk_size=min(512, cache_len),
                                    verbose=VerbosityLevels.NONE)
    out = list(batched_generate_fn(model=inf, prompts=ids, max_returned_tokens=8,
        sample_args=dict(temperature=1.0, top_k=1, top_p=1.0), stop_tokens=(), ignore_index=pad))
    return tok.decode(torch.cat(out, dim=1)[0])

print("prompt_len", L_, "answer", answer)
print("no-evict (cl=L+32):", repr(gen(L_ + 32)))   # -> '424242'
print("evict   (cl=1024) :", repr(gen(1024)))       # -> garbage (e.g. '\n\n\n...')
```

Expected: both should return `424242` (the last record is retained under
`lastrec` either way). Actual: the no-eviction case is correct; the eviction case
returns garbage.

**Environment**: Ubuntu 24.04, A10G, torch 2.13.0+cu130, `lastrec` / `h2o` caches,
eager SDPA (FlashInfer disabled to rule the kernel out). Reproduced on the
`grpo-cache-reuse` branch of `mmjerge/keys_values`.

**Related**: possibly the same underlying cause as #99 (`smart-lastrec`) and #5
(`h2o`/`qh2o` scoring); distinct from #139 (`ln_f`).

**Comment by mmjerge** [comment]
Narrowed this down considerably (branch `fix-evict-generation`):

**It's a mechanism bug, not lossiness — proven via RoPE relativity.** RoPE attention scores depend only on *relative* position, so the final-position attention over a retained suffix must be identical whether that suffix sits at positions `0..S-1` or `P..P+S-1`. Test on a 12-layer toy model (fp32, CPU):

- `f_dense = dense over [suffix]` (suffix at 0..S-1)
- `f_evict = lastrec(cl=S) over [prefix+suffix]` (evicts prefix, keeps suffix at P..P+S-1)

These should be equal; instead `max|Δ|` is ~0.15–0.22 (and argmax flips on deeper models).

**Exact trigger: partial writes into a full ring buffer** (i.e. a written chunk with `num != cache_length`). Sweeping `(P, S, cl, chunk)`:

| P | S | cl | chunk | writes | evict vs dense-suffix Δ |
|--:|--:|--:|--:|--|--:|
| 512 | 128 | 128 | 128 | all full-cl | **8e-7** ✓ |
| 384 | 128 | 128 | 128 | all full-cl | **1e-6** ✓ |
| 300 | 128 | 128 | 128 | last chunk = 44 (partial) | **0.21** ✗ |
| 128 | 128 | 128 | 32  | partial chunks | **0.19** ✗ |

When every write is exactly `cache_length` (clean full overwrite), eviction is correct. Any partial write corrupts the result.

**Implication for generation:** decode writes 1 token at a time (`num=1 << cache_length`), so **every decode step past cache-full is a partial write** — which is why autoregressive generation with any evicting cache (`lastrec`, `h2o`, quantized variants) produces garbage (immediate EOS). The first generated token is already wrong because the multi-chunk prefill's final chunk is typically partial.

The write path I inspected (`positions_wrap_around`, `DefaultKVCacheBuffers.set_slots`/`get_keys_values`, and `LastRecentlyInsertedKVCache._forward_internal`'s `token_pos` update) *looks* slot-aligned on paper, and `build_mask_slice` masks by `query_pos < token_positions` correctly — so the corruption is somewhere in how a partial/wrapped write's returned keys/values line up with `token_positions()` (and thus the mask) for the attention over the just-written chunk. Not yet pinned to a line, but the partial-write-vs-full-write repro above isolates it tightly.

Affects `lastrec` and `h2o` identically (not quantization). Likely the same root cause as #99 / #5.

**Comment by mmjerge** [comment]
**Correction to my previous comment — please disregard the "partial ring-buffer write" diagnosis. It was a flaw in my test, not a bug in the cache.**

My earlier comparison used "dense forward over the retained suffix only" as the reference. That equivalence is only valid in the special case where the retained set happens to be written as **one full-cache chunk** (then every layer attends only within the suffix). In the general case, the hidden states of retained tokens legitimately encode context that was later evicted, so a difference is *expected* lossiness.

Verification (`lastrec`, `init_grace_tokens=0`, cl=128, chunk=128, fp32):

| model | full-overwrite (L=640) vs dense(suffix) | partial/mixed (L=428) vs dense(suffix) |
|---|---:|---:|
| 1 layer | 7e-7 | **6e-7** |
| 12 layers | 8e-7 | **0.25** |

With **1 layer both cases match to ~1e-6**, so the write/slot/position/mask machinery is exact; the divergence only appears once hidden states can compound across layers. Direct instrumentation agrees: on a partial-write chunk the cached keys/values match a dense reference gathered at `token_positions()` **exactly (0.0)**, and the attention output matches a manual masked-softmax reference to **1.5e-7**.

**The actual cause of the garbage generation is attention-sink eviction.** Trained LLMs place large attention mass on the first few tokens (BOS / chat-template header) and use them as no-op sinks (Xiao et al., *Efficient Streaming Language Models with Attention Sinks*). Evicting them makes the softmax redistribute that mass onto content tokens, which collapses generation to immediate EOS. Toy random models have no sink structure, which is why unit tests pass while a real model fails.

Empirically, protecting the sinks/recent tokens fully restores quality (Qwen2.5-0.5B, A10G, ~2062-token prompts, evicting cache at cl=1024, answer in the last record):

| config | accuracy |
|---|---:|
| `lastrec`, `init_grace_tokens=0` (current default) | **0.00** |
| `lastrec`, `init_grace_tokens=16` | **0.90** |
| `lastrec`, `init_grace_tokens=64` | 0.90 |
| no eviction (cl > prompt) baseline | 0.90 |
| `h2o`, `grace_period=0` (current default) | **0.00** |
| `h2o`, `grace_period=64` | **0.70** |
| `h2o`, `keep_initial_fraction=0.10, grace_period=128` | 0.70 |

So this is a **default-configuration** problem rather than incorrect attention math: with `init_grace_tokens=0` / `grace_period=0`, any evicting cache produces garbage on real models, and those are the current defaults via `KVCacheFactory`.

Suggested fix (happy to open a PR): have `KVCacheFactory` inject safe defaults unless the caller specifies them — e.g. `init_grace_tokens = min(16, cache_length // 8)` for `lastrec`, and `grace_period = cache_length // 16` for the H2O family (`h2o-orig` excluded, it rejects `grace_period`). Explicit values, including `0`, stay respected so the raw mechanics can still be unit-tested. I have this implemented locally with the full `test/kvcache` suite passing (934 tests).

Note the `keep_initial_fraction` docstring already anticipates exactly this ("cater for initial prompt tokens which may be important"); it's the *default of 0* for the grace knobs that bites.

Apologies for the noise from the earlier incorrect analysis.

**Comment by mmjerge** [comment]
GPU verification of the proposed fix (safe eviction defaults), on a fresh g5.2xlarge (A10G), Qwen2.5-0.5B-Instruct, bf16, eager SDPA, ~2k-token prompts, n=40 per cell, substring exact-match accuracy:

| task | dense (full) | h2o @1024 | lastrec @512 | lastrec @1024 |
|---|---:|---:|---:|---:|
| recency (answer = last record) | 0.925 | **0.875** | **0.900** | **0.900** |
| needle_first | 0.700 | 0.200 | 0.000 | 0.000 |
| needle_last | 0.550 | 0.150 | 0.125 | 0.225 |

Before the fix, every evicting configuration scored **0.000 on all tasks** (immediate-EOS garbage). With sink/grace protection on, recency recovers to within noise of the dense baseline — i.e. generation through an evicting cache is functional again — while the needle rows now reflect genuine eviction lossiness rather than the collapse (the needle tokens really are evicted; that is the expected quality frontier of these policies).

One residual: `h2o` at a very tight budget (512 slots for ~2k context, grace 32) still collapses to 0; larger budgets behave. May be worth a floor on the post-grace budget or a documented minimum.

End-to-end GRPO training through the evicting cache also runs cleanly with the fix (generation + policy gradient, no inference-tensor issues), and with #141 applied the rollout-vs-training-forward log-prob skew drops from ~4.8 to ~0.13 (bf16 noise).

Happy to open a PR with the factory change (+ the measurements above) if the approach looks right to the maintainers — branch: `mmjerge/keys_values@grpo-upstream`.

**Comment by mseeger** [comment]
@mmjerge I am a bit confused by this, can you give a brief synopsis?

We use this code since quite a while with contexts > cache length, and we've not seen these issues.

I have results for using the pretrained model with our KV cache logics, and they are not really bad. They are of course worse than with fine-tuning, but they are not terrible.

Maybe you create an isolated example?

**Comment by mmjerge** [comment]
@mseeger You're right, and thanks for pushing back — I conflated two different mechanisms in my earlier synopsis. Let me separate them cleanly:

**1. `lastrec`: attention-sink eviction (your point, agreed).** `lastrec` evicts the *initial* tokens first, which are the sinks — hence the immediate-EOS garbage. `smart-lastrec` addressing exactly this makes sense, and a small `init_grace_tokens` default (you suggested 1 or a small number — our measurements say 16 fully restores a recency probe from 0.00 to 0.90, matching no-eviction) is the cheap fix for plain `lastrec`.

**2. H2O: you're correct that sinks are NOT the issue.** `keep_initial_fraction` preserves the initial tokens at the first eviction, and their accumulated scores keep them resident afterwards. The failure we measured for H2O is different: **freshly inserted tokens have the lowest accumulated scores**, so when eviction pressure is high, the most *recently read* region (typically the question at the end of the prompt) gets evicted before generation can use it. This is exactly what the `H2OKVCache` docstring anticipates ("summing scores over rounds may favor earlier tokens") and what `grace_period` exists for.

**Isolated example** (Qwen2.5-0.5B-Instruct, A10G, bf16, eager SDPA): ~2062-token prompts of key-value records where the answer is the value in the **last** record; `h2o-torch-quantized8`, `cache_length=1024` (so ~50% eviction pressure), chunked prefill (chunk 512), greedy decode, n=10, substring match:

| config | accuracy |
|---|---:|
| `grace_period=0` (current default) | **0.00** |
| `grace_period=64` | **0.70** |
| `keep_initial_fraction=0.10, grace_period=128` | 0.70 |
| no eviction (cl > prompt) | 0.90 |

Same prompts, same model — only the grace knob changes. Happy to package this as a runnable script if useful (it's `examples/longqa_task.py` + `longqa_quality.py` in PR #142, task `recency`).

**Why you likely haven't seen this**: I suspect three compounding differences vs your long-running evals: (a) much lower eviction pressure (e.g. 32k cache on 64k contexts vs our 0.5 ratio with the critical span at the very end), (b) fine-tuned models that have *adapted to the cache policy during training* (the policy sees evicted contexts in training, so it learns to cope — pre-trained checkpoints don't get that), and (c) tasks where the crucial information is distributed rather than concentrated in the recently-read tail. Our setting (pre-trained checkpoint, tight budget, answer at the end) maximizes the effect.

On the defaults question: I've updated the PR to state the two mechanisms separately in the factory comment. For `lastrec`, a small `init_grace_tokens` default seems uncontroversial. For H2O I'm happy either way — keep `grace_period=0` (no behavior change) and just document the tight-budget/recent-tail caveat, or default it to `cache_length//16` as in the PR. One honest limitation either way: at *very* tight budgets (512 slots for a 2k prompt) H2O still collapses even with grace — there may be a practical floor on the post-grace budget worth documenting.

**Comment by mseeger** [comment]
Hi, about H2O: What you mention in 2, is also dealt with by using `normalize_scores=True`, I think. This is why I'd leave the `grace_period` default at 0 here.

In the results for the paper, indeed `normalize_scores=True` often works better than default (but not always). However, also the original H2O variant (which averages scores over the batch dimension) also often works better than the default.

Let us keep this simple and just change the default for `lastrec`, citing attention sinks as the reason.

Of course, once you have a real application, such as RL, you want to optimize the configuration. But the defaults should be simple, unless we have strong evidence otherwise.

**Comment by mmjerge** [comment]
Sounds good — done in PR #142:

- Factory default now **only** for `lastrec` (`init_grace_tokens = min(16, cache_length//8)`), citing attention sinks; the H2O family's `grace_period` stays 0, no behavior change.
- The `grace_period` my RL experiments rely on moved into the example configs explicitly (application-level tuning, as you suggest), and the factory comment points to `normalize_scores=True` as the alternative mitigation for the recent-tail effect.

Thanks — agreed that simple defaults win absent strong evidence. The tight-budget H2O measurements stay documented here for anyone who hits that regime.

**Comment by mseeger** [comment]
Closing this. Should be resolved by way of #146
